sdk/go/common/slice/slice.go

```diff
@@ -0,0 +1,25 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slice
+
+// Preallocates a slice of type T of a given length. If the input length is 0 then the returned slice will be
+// nil. We prefer this over `make([]T, 0, length)` because nil is often treated semantically different from an
+// empty slice.
+func Prealloc[T any](length int) []T {
+	if length == 0 {
+		return nil
+	}
+	return make([]T, length)
```

Comment:
```suggestion
	return make([]T, 0, length)
```

Reply:
Ah yup, that would have triggered a bucket load of tests to fail 😆